src/System.Configuration/src/System/Configuration/ClientConfigPaths.cs

```diff
@@ -0,0 +1,359 @@
+                RoamingConfigFilename = CombineIfValid(RoamingConfigDirectory, UserConfigFilename);
+            }
+
+            string localFolderPath = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData);
+            if (Path.IsPathRooted(localFolderPath))
+            {
+                LocalConfigDirectory = CombineIfValid(localFolderPath, dirSuffix);
+                LocalConfigFilename = CombineIfValid(LocalConfigDirectory, UserConfigFilename);
+            }
+        }
+
+        internal static ClientConfigPaths Current => GetPaths(null, true);
+
+        internal bool HasEntryAssembly { get; }
+
+        internal string ApplicationUri { get; }
+
+        internal string ApplicationConfigUri { get; }
+
+        internal string RoamingConfigFilename { get; }
+
+        internal string RoamingConfigDirectory { get; }
+
+        internal bool HasRoamingConfig => (RoamingConfigFilename != null) || !_includesUserConfig;
+
+        internal string LocalConfigFilename { get; }
+
+        internal string LocalConfigDirectory { get; }
+
+        internal bool HasLocalConfig => (LocalConfigFilename != null) || !_includesUserConfig;
+
+        internal string ProductName { get; private set; }
+
+        internal string ProductVersion { get; private set; }
+
+        internal static ClientConfigPaths GetPaths(string exePath, bool includeUserConfig)
+        {
+            ClientConfigPaths result;
+
+            if (exePath == null)
+            {
+                if ((s_current == null) || (includeUserConfig && !s_currentIncludesUserConfig))
+                {
+                    s_current = new ClientConfigPaths(null, includeUserConfig);
+                    s_currentIncludesUserConfig = includeUserConfig;
+                }
+
+                result = s_current;
+            }
+            else result = new ClientConfigPaths(exePath, includeUserConfig);
+
+            return result;
+        }
+
+        internal static void RefreshCurrent()
+        {
+            s_currentIncludesUserConfig = false;
+            s_current = null;
+        }
+
+        // Combines path2 with path1 if possible, else returns null.
+        private static string CombineIfValid(string path1, string path2)
+        {
+            string returnPath = null;
+
+            if ((path1 == null) || (path2 == null)) return null;
+
+            try
+            {
+                string combinedPath = Path.Combine(path1, path2);
+                if (combinedPath.Length < MaxPath) returnPath = combinedPath;
+            }
+            catch { }
+
+            return returnPath;
+        }
+
+        // Returns a type and hash suffix based on app domain evidence. The evidence we use, in
+        // priority order, is Strong Name, Url and Exe Path. If one of these is found, we compute a 
+        // SHA1 hash of it and return a suffix based on that. If none is found, we return null.
```

Comment:
To be clear- it does the same thing, but using what "used to come from" app domain evidence. This is the same thing I had to do in IsolatedStorage. I'm following up with another change to move the relevant code to common.